src/Controller/CommonAjaxController.php

```diff
@@ -0,0 +1,363 @@
+        if ($error_response instanceof Response) {
+            return $error_response;
+        }
+
+        if ($this->item->restore($post)) {
+            // Successfull restoration
+            // Feedback message is already handled by CommonDBTM::addMessageOnRestoreAction
+            return $this->successResponse(200, [
+                "is_deleted" => false,
+            ]);
+        } else {
+            // Failed to restore
+            $error = $this->item->formatSessionMessageAfterAction(
+                __("Failed to restore item")
+            );
+            return $this->errorReponse(422, $error);
+        }
+    }
+
+    /**
+     * Handle "purge" action
+     *
+     * @param array $post POST data
+     *
+     * @return Response
+     */
+    final protected function handlePurgeAction(array $post): Response
+    {
+        // Validate rights
+        $error_response = $this->check((int) $post['id'], PURGE, $post);
+        if ($error_response instanceof Response) {
+            return $error_response;
+        }
+
+        if ($this->item->delete($post, true)) {
+            // Successfull purge
+            // Feedback message is already handled by CommonDBTM::addMessageOnPurgeAction
+            return $this->successResponse(200, [
+                "redirect" => $this->item->getSearchURL(),
+            ]);
+        } else {
+            // Failed purge
+            $error = $this->item->formatSessionMessageAfterAction(
+                __("Failed to purge item")
+            );
+            return $this->errorReponse(422, $error);
+        }
+    }
+
+    /**
+     * Similar checks than done in CommonDBTM::check() without parts that aren't
+     * compatible with an AJAX response
+     *
+     * If a check fail, an error reponse will be returned
+     * If everything is OK, `null` is returned
+     *
+     * @param int   $id     Target item's id
+     * @param int   $right  Specific right to check
+     * @param array $post   POST data
+     *
+     * @return Response|null
+     */
+    final protected function check(int $id, int $right, array $post): ?Response
+    {
+        if (!$this->item->checkIfExistOrNew($id)) {
+            return $this->errorReponse(403, __("Item not found"));
+        } elseif (!$this->item->can($id, $right, $input)) {
+            return $this->errorReponse(
+                403,
+                __("You don't have permission to perform this action.")
+            );
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Check that the given class is whitelisted for ajax updates
+     * This is needed as we don't want others controllers to be bypassed by this
+     * endpoint
+     *
+     * @param string $class Given class
+     *
+     * @return bool
+     */
+    final protected function isClassWhitelisted(string $class): bool
+    {
+        return in_array($class, CommonAjaxControllerWhitelist::getClasses());
+    }
```

Comment:
Please try to avoid `master`/`slave` and `white`/`black` words in new code.

```suggestion
    final protected function isClassAllowed(string $class): bool
    {
        return in_array($class, CommonAjaxControllerAllowlist::getClasses());
    }
```